File: src/main/java/local/queens/State.java
package local.queens;

import java.util.Arrays;
import java.util.Iterator;
import java.util.NoSuchElementException;
import java.util.Random;

import core.Action;

/** Represents a single state in the 8 Queens Problem (generalised to the X Queens Problem). */
public class State implements core.State<State> {
	/** Represents the chess board. Every position in the array corresponds to a column on the board, and the integer in that position indicates the row the queen is in. */
	private int[] board;
	
	private Double qualityScore;

	/**
	 * Creates a valid state with the specified dimensions and random queen placement.
	 * @param dimensions The number of rows and columns the board should have.
	 */
	public State(int dimensions) {
		board = new int[dimensions];
		Random rng = new Random();
		for (int column = 0; column < dimensions; column++) {
			// Place the queen for this column in a randomly selected row.
			board[column] = rng.nextInt(dimensions);
		}
	}
	
	/**
	 * Creates a state based using the provided board configuration.
	 * @param board The board that the state should represent.
	 */
	public State(int[] board) {
		this.board = board;
	}

	@Override
	public boolean isGoalState() {
		// Loop over the columns on the board. 
		for (int columnBeingChecked = 0; columnBeingChecked < board.length; columnBeingChecked++) {
			// Loop over the columns to the right of the current column.
			for (int columnToCheckAgainst = columnBeingChecked + 1; columnToCheckAgainst < board.length; columnToCheckAgainst++) {
				if (queensAreClashing(columnBeingChecked, columnToCheckAgainst)) {
					// These two queens are clashing. This board is not a solution.
					return false;
				}
			}
		}
		
		// All checks succeeded.
		return true;
	}
	
	/**
	 * Checks whether the two indicated queens are clashing.
	 * @param column1 The column containing one of the queens.
	 * @param column2 The column containing the other queen.
	 * @return {@code true} if the two indicated queens are in such a relative position that they threaten each other.
	 */
	public boolean queensAreClashing(int column1, int column2) {
		// Calculate the horizontal and vertical distances between the queens in the two columns.
		int horizontalDistance = Math.abs(column1 - column2);
		int verticalDistance = Math.abs(board[column1] - board[column2]);
		
		// The queens are clashing if
		//   - the vertical distance is zero (which means that the two queens are in the same row, clashing horizontally), OR
		//   - the vertical distance is equal to the horizontal distance (which means that the queens are clashing diagonally).
		// Thanks to the data structure, this method does not need to check for vertical clashing.
		return verticalDistance == 0 || horizontalDistance == verticalDistance;
	}
	
	/**
	 * Creates an {@link Iterator} that provides all states 'reachable' from this one. In this case, a 'reachable' state is a state produced by moving a single queen to another
	 * position in the same column.
	 */
	public Iterator<Action<State>> createAvailableActionsIterator() {
		return new Iterator<Action<State>>() {
			// The column that the cursor is currently on.
			private int column;
			// The row that the cursor is currently on.
			private int row = -1;
			
			@Override
			public boolean hasNext() {
				// Check if there is an open position after the cursor.
				if (column == board.length - 1
						&& row == board.length - 2
						&& board[column] == board.length - 1) {
					// This is a special case: the queen in the last column is also in the last row. The cursor is on the position immediately preceding this queen (last column,
					// second-to-last row) so no more open positions are available.
					return false;
				}
				
				// In general, another successor state is available if the cursor is not on the very last space of the board.
				return column < board.length - 1 || row < board.length - 1;
			}
			
			@Override
			public Action<State> next() {
				if (!hasNext()) {
					throw new NoSuchElementException("No next successor available.");
				}
				
				// Find the next open position on the board.
				findNextOpenPosition();
				// Move the queen in the current column to the open position, which will result in a different board configuration. Create a new action for it and add it to the
				// list.
				int[] successorBoard = board.clone();
				successorBoard[column] = row;
				return new Action<>(new State(successorBoard), 0);
			}
			
			/** Moves the cursor to the next open position on the board. */
			private void findNextOpenPosition() {
				do {
					nextPosition();
					// Keep going until the current position does not contain an queen.
				} while (board[column] == row);
			}
			
			/** Moves the cursor to the next position on the board. */
			private void nextPosition() {
				row++;
				if (row >= board.length) {
					row = 0;
					column++;
				}
			}
		};
	}
	
	@Override
	public Action<State> randomlySelectAvailableAction() {
		Random rng = new Random();
		
		// Randomly select a queen to move.
		int column = rng.nextInt(board.length);
		// Randomly select a row to move her to.
		int row = rng.nextInt(board.length - 1);
		if (row >= board[column]) {
			// The selected row is equal to or greater than the row the selected queen is in. Move down one row to compensate for this.
			row++;
		}
		
		// Clone the board and move the queen to her new position on this board.
		int[] successorBoard = board.clone();
		successorBoard[column] = row;
		return new Action<>(new State(successorBoard), 0);
	}
	
	@Override
	public boolean equals(Object otherObject) {
		if (super.equals(otherObject)) {
			// This is the same state object.
			return true;
		}
		
		if (!(otherObject instanceof State)) {
			// The other object is not of the same class.
			return false;
		}
		
		// The other object is an instance of this class. The only thing that matters now is the board configuration.
		return Arrays.equals(this.board, ((State) otherObject).board);
	}
	
	@Override
	public int hashCode() {
		return Arrays.hashCode(board);
	}
	
	@Override
	public String toString() {
		// Prints the board [0,2,1] as follows:
		// [()    ]
		// [    ()]
		// [  ()  ]
		// This is done to give a close-to-square image of the board, for easier viewing.
		
		StringBuilder result = new StringBuilder();
		for (int row = 0; row < board.length; row++) {
			result.append("[");
			
			for (int column : board) {
				result.append(column == row ? "()" : "  ");
			}
			
			result.append("]");
			
			if (row < board.length - 1) {
				result.append("\n");
			}
		}
		
		if (qualityScore != null) {
			result.append("\nQuality: ").append(qualityScore);
		}
		
		return result.toString();
	}
	
	/** @return The length of one side of the board. Because a board in this problem is always square, this length describes the size in both dimensions. */
	public int determineDimensions() {
		return board.length;
	}

	@Override
	public Double getQualityScore() {
		return qualityScore;
	}
	
	@Override
	public void setQualityScore(Double qualityScore) {
		this.qualityScore = qualityScore;
	}
}

File: src/main/scala/nl/calmamity/search/local/queens/State.scala
package nl.calmamity.search.local.queens

import java.util

import core.Action

import nl.calmamity.search.core

object State {
	val randomNumbers = new scala.util.Random()

	/**
	  * Creates a valid state with the specified dimensions and random queen placement.
	  * @param numDimensions The number of rows and columns the board should have.
	  */
	def apply(numDimensions: Int): State = {
		val board = for (_ <- 1 to numDimensions) yield randomNumbers.nextInt(numDimensions)
		State(board)
	}
}

/**
  * Represents a single state in the 8 Queens Problem (generalised to the X Queens Problem).
  * @param board Represents the chess board. Every position in the array corresponds to a column on the board, and the
  * integer in that position indicates the row the queen is in.
  */
case class State(board: Seq[Int]) extends core.State[State] {
	override def isGoalState: Boolean = {
		val pairsToCheck = for {
			columnBeingChecked <- board.indices
			columnToCheckAgainst <- columnBeingChecked + 1 until board.size
		} yield (columnBeingChecked, columnToCheckAgainst)

		!pairsToCheck.exists {
			case (columnBeingChecked, columnToCheckAgainst) =>
				queensAreClashing(columnBeingChecked, columnToCheckAgainst)
		}
	}

	/**
	  * Checks whether the two indicated queens are clashing.
	  * @param column1 The column containing one of the queens.
	  * @param column2 The column containing the other queen.
	  * @return `true` if the two indicated queens are in a position to threaten each other, `false` otherwise.
	  */
	def queensAreClashing(column1: Int, column2: Int): Boolean = {
		// Calculate the horizontal and vertical distances between the queens in the two columns.
		val horizontalDistance = math.abs(column1 - column2)
		val verticalDistance = math.abs(board(column1) - board(column2))

		// The queens are clashing if
		//   - the vertical distance is zero (which means that the two queens are in the same row, clashing
		//     horizontally), OR
		//   - the vertical distance is equal to the horizontal distance (which means that the queens are clashing
		//     diagonally).
		// Thanks to the data structure, this method does not need to check for vertical clashing.
		verticalDistance == 0 || horizontalDistance == verticalDistance
	}

	/**
	  * Creates an [[Iterator]] that provides all states that can be created by moving a single queen to another
	  * position in the same column.
	  */
	override def createAvailableActionsIterator(): Iterator[Action[State]] = new Iterator[Action[State]]() {
		// The column that the cursor is currently on.
		var column: Int = 0
		// The row that the cursor is currently on.
		var row: Int = -1

		override def hasNext(): Boolean = {
			// Check if there is an open position after the cursor.
			if (column == board.size - 1
					&& row == board.size - 2
					&& board(column) == board.size - 1) {
				// This is a special case: the queen in the last column is also in the last row. The cursor is on
				// the position immediately preceding this queen (last column, second-to-last row) so no more open
				// positions are available.
				false
			} else {
				// In general, another successor state is available if the cursor is not on the very last space of
				// the board.
				column < board.length - 1 || row < board.length - 1
			}
		}

		override def next(): Action[State] = {
			if (!hasNext()) {
				throw new NoSuchElementException("No next successor available.")
			}

			// Find the next open position on the board.
			moveCursorToNextOpenPosition()
			// Move the queen in the current column to the open position, which will result in a different board
			// configuration. Create a new action for it and add it to the list.
			val successorBoard = board.updated(column, row)
			new Action[State](new State(successorBoard), 0)
		}

		/** Moves the cursor to the next open position on the board. */
		def moveCursorToNextOpenPosition(): Unit = {
			do {
				moveCursorToNextPosition()
				// Keep going until the current position does not contain an queen.
			} while (board(column) == row)
		}

		/** Moves the cursor to the next position on the board. */
		def moveCursorToNextPosition(): Unit = {
			row += 1
			if (row >= board.length) {
				row = 0
				column += 1
			}
		}
	}

	override def randomlySelectAvailableAction: Action[State] = {
		// Randomly select a queen to move.
		val column = State.randomNumbers.nextInt(board.size)
		// Randomly select a row to move her to.
		val row = State.randomNumbers.nextInt(board.length - 1) match {
			case value if value < board(column) =>
				value
			case value =>
				// The selected row is equal to or greater than the row the selected queen is in. Move down one row to
				// compensate for this.
				value + 1
		}

		// Clone the board and move the queen to her new position on this board.
		val successorBoard = board.updated(column, row)
		new Action[State](new State(successorBoard), 0)
	}

	override def equals(other: scala.Any): Boolean = {
		other match {
			case otherState: State =>
				this.board == otherState.board
		}
	}

	override def hashCode(): Int = util.Arrays.hashCode(board.toArray)

	override def toString: String = {
		// Prints the board [0,2,1] as follows:
		// [()    ]
		// [    ()]
		// [  ()  ]
		// This is done to give a close-to-square image of the board, for easier viewing.

		val rowsAsStrings = for (row <- board.indices) yield
			board.map {
				column =>
					if (column == row) {
						"()"
					} else {
						"  "
					}
			}
			.mkString("[", "", "]")

		rowsAsStrings.mkString("\n")
	}

	/**
	  * @return The length of one side of the board. Because a board in this problem is always square, this length
	  * describes the size in both dimensions.
	  */
	def determineDimensions: Int = {
		board.length
	}
}
